File: backend/api/routers/catalog.py
from api.core.deps import check_admin, get_catalog_repo, get_user_repo
from api.core.ml import get_books_with_ml
from api.repos.catalog import CatalogRepo
from api.repos.users import UserRepo
from api.schemas.catalog import (
    AuthorCreate,
    AuthorInfo,
    AuthorRead,
    AuthorUpdate,
    BindingRead,
    BookCreate,
    BookFilter,
    BookInfo,
    BookRead,
    BookUpdate,
    CatCreate,
    CatRead,
    PubCreate,
    PubRead,
    PubUpdate,
    TopicCreate,
    TopicRead,
)
from api.schemas.users import ReviewFilter, ReviewRead
from fastapi import APIRouter, Depends, Query, Request, Response
from typing_extensions import Annotated

router = APIRouter(prefix="/catalog", tags=["Catalog"])


@router.get("/books", response_model=list[BookRead])
async def get_books_by_filter(
    filter: Annotated[BookFilter, Query()], repo: CatalogRepo = Depends(get_catalog_repo)
) -> list[BookRead]:
    return await repo.get_books(filter)


@router.get("/books/by_query", response_model=list[BookRead])
async def get_books_by_query(books: BookRead = Depends(get_books_with_ml)):
    return books


@router.get("/books/{id}", response_model=BookInfo)
async def get_book_by_id(
    request: Request, response: Response, id: int, repo: CatalogRepo = Depends(get_catalog_repo)
) -> BookInfo:
    book = await repo.get_book_by_id(id)
    viewed_raw = request.cookies.get("viewed_books", "")
    viewed_books = [int(x) for x in viewed_raw.split(",") if x.isdigit()]
    if book.id not in viewed_books:
        await repo.update_book_views_count(id)
        viewed_raw += f",{id}"
        response.set_cookie(
            "viewed_books", viewed_raw, httponly=True, samesite="lax", max_age=3600 * 24
        )
    return book


@router.post("/books")
async def add_book(
    book: BookCreate, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    id = await repo.create_book(book)
    return {"id": id, "status": "success"}


@router.patch("/books/{id}")
async def update_book(
    id: int,
    data: BookUpdate,
    admin=Depends(check_admin),
    repo: CatalogRepo = Depends(get_catalog_repo),
) -> dict:
    print(data)
    await repo.update_book(id, data)
    return {"status": "success"}


@router.delete("/books/{id}", response_model=dict[str, str])
async def delete_book(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict[str, str]:
    await repo.delete_book(id)
    return {"status": "success"}


@router.get("/books/{id}/reviews", response_model=list[ReviewRead])
async def get_book_reviews(
    id: int, filter: Annotated[ReviewFilter, Query()], repo: UserRepo = Depends(get_user_repo)
) -> list[ReviewRead]:
    return await repo.get_reviews(filter, book_id=id, with_text=True)


@router.get("/cats", response_model=list[CatRead])
async def get_cats(repo: CatalogRepo = Depends(get_catalog_repo)) -> list[CatRead]:
    return await repo.get_cats()


@router.post("/cats")
async def add_cat(
    cat: CatCreate, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    id = await repo.create_cat(cat)
    return {"id": id, "status": "success"}


@router.delete("/cats/{id}")
async def delete_cat(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    await repo.delete_cat(id)
    return {"status": "success"}


@router.get("/cats/{cat_id}/topics", response_model=list[TopicRead])
async def get_cat_topics(
    cat_id: int, repo: CatalogRepo = Depends(get_catalog_repo)
) -> list[TopicRead]:
    return await repo.get_topics(cat_id)


@router.get("/cats/{cat_id}/authors", response_model=list[AuthorRead])
async def get_cat_authors(
    cat_id: int, repo: CatalogRepo = Depends(get_catalog_repo)
) -> list[AuthorRead]:
    return await repo.get_authors(cat_id)


@router.post("/cats/{cat_id}/topics/{topic_id}")
async def link_topic(
    cat_id: int,
    topic_id: int,
    admin=Depends(check_admin),
    repo: CatalogRepo = Depends(get_catalog_repo),
) -> dict:
    await repo.link_topic(topic_id, cat_id)
    return {"status": "success"}


@router.delete("/cats/{cat_id}/topics/{topic_id}")
async def delink_topic(
    cat_id: int,
    topic_id: int,
    admin=Depends(check_admin),
    repo: CatalogRepo = Depends(get_catalog_repo),
) -> dict:
    await repo.delink_topic(topic_id, cat_id)
    return {"status": "success"}


@router.get("/topics", response_model=list[TopicRead])
async def get_topics(repo: CatalogRepo = Depends(get_catalog_repo)) -> list[TopicRead]:
    return await repo.get_topics()


@router.post("/topics")
async def add_topic(
    topic: TopicCreate, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    id = await repo.create_topic(topic)
    return {"id": id, "status": "success"}


@router.delete("/topics/{id}")
async def delete_topic(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    await repo.delete_topic(id)
    return {"status": "success"}


@router.get("/pubs", response_model=list[PubRead])
async def get_pubs(repo: CatalogRepo = Depends(get_catalog_repo)) -> list[TopicRead]:
    return await repo.get_pubs()


@router.get("/pubs/{id}")
async def get_pub(id: int, repo: CatalogRepo = Depends(get_catalog_repo)) -> AuthorInfo:
    return await repo.get_pub_by_id(id)


@router.post("/pubs")
async def add_pub(
    pub: PubCreate, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    id = await repo.create_pub(pub)
    return {"id": id, "status": "success"}


@router.patch("/pubs/{id}")
async def update_pub(
    id: int,
    data: PubUpdate,
    admin=Depends(check_admin),
    repo: CatalogRepo = Depends(get_catalog_repo),
) -> dict:
    await repo.update_pub(id, data)
    return {"status": "success"}


@router.delete("/pubs/{id}")
async def delete_pub(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    await repo.delete_pub(id)
    return {"status": "success"}


@router.get("/authors", response_model=list[AuthorRead])
async def get_authors(repo: CatalogRepo = Depends(get_catalog_repo)) -> list[AuthorRead]:
    return await repo.get_authors()


@router.get("/authors/{id}")
async def get_author(id: int, repo: CatalogRepo = Depends(get_catalog_repo)):
    return await repo.get_author_by_id(id)


@router.post("/authors")
async def add_author(
    author: AuthorCreate, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    id = await repo.create_author(author)
    return {"id": id, "status": "success"}


@router.patch("/authors/{id}")
async def update_author(
    id: int,
    data: AuthorUpdate,
    admin=Depends(check_admin),
    repo: CatalogRepo = Depends(get_catalog_repo),
):
    await repo.update_author(id, data)
    return {"status": "success"}


@router.delete("/authors/{id}")
async def delete_author(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    await repo.delete_author(id)
    return {"status": "success"}


@router.get("/bindings", response_model=list[BindingRead])
async def get_bindings(repo: CatalogRepo = Depends(get_catalog_repo)) -> list[str]:
    return await repo.get_bindings()


@router.post("/bindings")
async def add_binding(
    name: str, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
) -> dict:
    await repo.add_binding()
    return {"status": "success"}


@router.delete("/bindings/{id}")
async def delete_binding(
    id: int, admin=Depends(check_admin), repo: CatalogRepo = Depends(get_catalog_repo)
):
    await repo.delete_binding(id)
    return {"status": "success"}

File: frontend/src/api/catalog.ts
import { apiFetch } from "./client";
import type {
  BookRead,
  BookInfo,
  BookFilter,
  CatRead,
  TopicRead,
  AuthorRead,
  PubRead,
  BookCreate,
  BookUpdate,
  CatCreate,
  PubCreate,
  PubUpdate,
  PubInfo,
  AuthorCreate,
  AuthorUpdate,
  AuthorInfo,
  BindingRead,
} from "../types/catalog";
import type { ReviewFilter, ReviewRead } from "../types/users";

const BOOKS_REL = "/catalog/books";
const AUTHORS_REL = "/catalog/authors";
const CATS_REL = "/catalog/cats";
const PUBS_REL = "/catalog/pubs";
const TOPICS_REL = "/catalog/topics";
const BINDING_REL = "/catalog/bindings";

export const catalogApi = {
  // GET /catalog/books
  async getBooks(filters: Partial<BookFilter> = {}): Promise<BookRead[]> {
    const params = new URLSearchParams();

    // Одиночные значения
    if (filters.cat_id !== null && filters.cat_id !== undefined)
      params.append("cat_id", String(filters.cat_id));
    if (filters.min_price !== null && filters.min_price !== undefined)
      params.append("min_price", String(filters.min_price));
    if (filters.max_price !== null && filters.max_price !== undefined)
      params.append("max_price", String(filters.max_price));
    if (filters.is_new !== null && filters.is_new !== undefined)
      params.append("is_new", String(filters.is_new));
    if (filters.is_bestseller !== null && filters.is_bestseller !== undefined)
      params.append("is_bestseller", String(filters.is_bestseller));
    if (filters.min_rating !== null && filters.min_rating !== undefined)
      params.append("min_rating", String(filters.min_rating));
    if (filters.search) params.append("search", filters.search);

    // Пагинация и сортировка
    if (filters.limit) params.append("limit", String(filters.limit));
    if (filters.offset !== null && filters.offset !== undefined)
      params.append("offset", String(filters.offset));
    if (filters.sort_by) params.append("sort_by", filters.sort_by);
    if (filters.reverse !== null && filters.reverse !== undefined)
      params.append("reverse", String(filters.reverse));

    // Массивы (list[int] в FastAPI передаются через повторение ключа)
    if (filters.topic_id && filters.topic_id.length > 0) {
      filters.topic_id.forEach((id) => params.append("topic_id", String(id)));
    }
    if (filters.binding_id && filters.binding_id.length > 0) {
      filters.binding_id.forEach((id) =>
        params.append("binding_id", String(id)),
      );
    }
    if (filters.author_id && filters.author_id.length > 0) {
      filters.author_id.forEach((id) => params.append("author_id", String(id)));
    }
    if (filters.pub_id && filters.pub_id.length > 0) {
      filters.pub_id.forEach((id) => params.append("pub_id", String(id)));
    }

    return apiFetch<BookRead[]>(`${BOOKS_REL}?${params.toString()}`);
  },

  async addBook(book: any): Promise<{ id: number; status: string }> {
    return apiFetch(BOOKS_REL, {
      method: "POST",
      body: JSON.stringify(book),
    });
  },
  // GET /catalog/books/{id}
  async getBookById(id: number): Promise<BookInfo> {
    return apiFetch<BookInfo>(`${BOOKS_REL}/${id}`);
  },

  async updateBook(id: number, data: BookUpdate): Promise<{ status: string }> {
    return apiFetch(`${BOOKS_REL}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    });
  },

  async deleteBook(id: number): Promise<{ status: string }> {
    return apiFetch(`${BOOKS_REL}/${id}`, { method: "DELETE" });
  },

  async getBookReviews(
    book_id: number,
    filter: Partial<ReviewFilter> = {},
  ): Promise<ReviewRead[]> {
    const params = new URLSearchParams();
    if (filter.user_id !== null && filter.user_id !== undefined)
      params.append("user_id", String(filter.user_id));
    if (filter.bad_first !== null && filter.bad_first !== undefined)
      params.append("bad_first", String(filter.bad_first));
    else if (filter.good_first !== null && filter.good_first !== undefined)
      params.append("good_first", String(filter.good_first));
    if (filter.limit !== null && filter.limit !== undefined)
      params.append("limit", String(filter.limit));
    if (filter.offset !== null && filter.offset !== undefined)
      params.append("offset", String(filter.offset));
    return apiFetch<ReviewRead[]>(`${BOOKS_REL}/${book_id}/reviews`);
  },

  async getCats(): Promise<CatRead[]> {
    return apiFetch<CatRead[]>("/catalog/cats");
  },

  async addCat(cat: CatCreate): Promise<{ status: string }> {
    return apiFetch(`${CATS_REL}`, {
      method: "POST",
      body: JSON.stringify(cat),
    });
  },

  async deleteCat(cat_id: number): Promise<{ status: string }> {
    return apiFetch(`${CATS_REL}/${cat_id}`, { method: "DELETE" });
  },

  async getTopics(cat_id: number | null = null): Promise<TopicRead[]> {
    if (cat_id !== null && cat_id !== undefined) {
      return apiFetch<TopicRead[]>(`${CATS_REL}/${cat_id}/topics`);
    } else {
      return apiFetch<TopicRead[]>(`${TOPICS_REL}`);
    }
  },

  async getAuthors(cat_id: number | null = null): Promise<AuthorRead[]> {
    if (cat_id !== null && cat_id !== undefined)
      return apiFetch<AuthorRead[]>(`${CATS_REL}/${cat_id}/authors/`);
    else return apiFetch<AuthorRead[]>(`${AUTHORS_REL}`);
  },
  async getPubs(): Promise<PubRead[]> {
    return apiFetch<PubRead[]>(PUBS_REL);
  },

  async addPub(pub: PubCreate): Promise<{ status: string }> {
    return apiFetch(PUBS_REL, { method: "POST", body: JSON.stringify(pub) });
  },

  async updatePub(
    pub_id: number,
    data: PubUpdate,
  ): Promise<{ status: string }> {
    return apiFetch(`${PUBS_REL}/${pub_id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    });
  },

  async getPub(pub_id: number): Promise<PubInfo> {
    return apiFetch<PubInfo>(`${PUBS_REL}/${pub_id}`);
  },

  async deletePub(pub_id: number): Promise<{ status: string }> {
    return apiFetch(`${PUBS_REL}/${pub_id}`, { method: "DELETE" });
  },

  async getAuthor(id: number): Promise<AuthorInfo> {
    return apiFetch<AuthorInfo>(`${AUTHORS_REL}/${id}`);
  },

  async addAuthor(author: AuthorCreate): Promise<{ status: string }> {
    return apiFetch(AUTHORS_REL, {
      method: "POST",
      body: JSON.stringify(author),
    });
  },

  async updateAuthor(
    id: number,
    data: AuthorUpdate,
  ): Promise<{ status: string }> {
    return apiFetch(`${AUTHORS_REL}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    });
  },

  async deleteAuthor(id: number): Promise<{ status: string }> {
    return apiFetch(`${AUTHORS_REL}/${id}`, { method: "DELETE" });
  },

  async getBindings(): Promise<BindingRead[]> {
    return apiFetch<BindingRead[]>(BINDING_REL);
  },

  async addBinding(name: string): Promise<{ status: string }> {
    return apiFetch(BINDING_REL, {
      method: "POST",
      body: JSON.stringify(name),
    });
  },

  async deleteBindind(id: number): Promise<{ status: string }> {
    return apiFetch(`${BINDING_REL}/${id}`, { method: "DELETE" });
  },
  async uploadBookCover(
    bookId: number,
    file: File,
  ): Promise<{ picture_url: string }> {
    const formData = new FormData();
    formData.append("file", file);

    return apiFetch<{ picture_url: string }>(`/media/catalog/books/${bookId}`, {
      method: "PATCH",
      body: formData,
    });
  },
  async getBooksByQuery(q: string): Promise<BookRead[]> {
    return apiFetch<BookRead[]>(
      `${BOOKS_REL}/by_query?q=${encodeURIComponent(q)}`,
    );
  },
};
